### DISCUSSION TITLE
Install packages declaratively

### DISCUSSION THREAD

**Discussion by chuckwondo** [twpayne/chezmoi#4057] [Q&A] [answered]
I'm new to chezmoi (and excited to get it working), and I'm trying to get started with chezmoi by installing my brew packages declaratively, as described in the example here: https://www.chezmoi.io/user-guide/advanced/install-packages-declaratively/

However, I'm getting the following error:

```plain
chezmoi: darwin-install-packages.sh: template: run_onchange_darwin-install-packages.sh.tmpl:5:18: executing "run_onchange_darwin-install-packages.sh.tmpl" at <.packages.darwin.brews>: map has no entry for key "packages"
```

Perhaps this is due to my having difficulty understanding the correct procedure for setting up my files with chezmoi, so please provide some guidance if the steps shown below are not how I should be doing things. I'm obviously missing something.

I do find it confusing to know exactly where to put files and when to literally place them in the paths shown in the code snippet headings and when to place them in my home dir and run `chezmoi add` instead.

Here's are the steps I took that led to the error above:

1. Created the directory `~/.chezmoidata`
2. Created the file `~/.chezmoidata/packages.yaml` (see below)
3. Ran `chezmoi add ~/.chezmoidata`
4. Ran `chezmoi cd`
5. Ran `$EDITOR run_onchange_darwin-install-packages.sh.tmpl`
6. Copy-pasted the contents from the linked example and saved my changes (see below)
7. Ran `exit` (to back out of `chezmoi cd`)
8. Ran `chezmoi apply`, which produced the error shown above (and so does `chezmoi diff`)

Here is what `chezmoi managed` shows:

```plain
.chezmoidata
.chezmoidata/packages.yaml
darwin-install-packages.sh
```

Here is my minimal `~/.chezmoidata/packages.yaml` file:

```yaml
packages:
  darwin:
    brews:
      - gh
    casks:
      - brave-browser
```

Here is my `~/.local/share/chezmoi/run_onchange_darwin-install-packages.sh.tmpl`:

```plain
{{ if eq .chezmoi.os "darwin" -}}
#!/bin/bash

brew bundle --no-lock --file=/dev/stdin <<EOF
{{ range .packages.darwin.brews -}}
brew {{ . | quote }}
{{ end -}}
{{ range .packages.darwin.casks -}}
cask {{ . | quote }}
{{ end -}}
EOF
{{ end -}}
```

**Comment by bradenhilton** [comment, marked as answer]
`.chezmoidata/` needs to be in your source directory (`~/.local/share/chezmoi` by default).

**Reply by chuckwondo** [reply]
I just discovered that. You beat me to it :-)

However, this is where my confusion comes in. Is there somewhere in the documentation that clarifies this point?

As I mentioned, I'm confused as to when I should explicitly (manually) place files directly under `~/.local/share/chezmoi` and when I should place them in my home dir and let `chezmoi add` implicitly (automatically) copy them to `~/.local/share/chezmoi`.

After a bit more digging, it appears that all [special files](https://www.chezmoi.io/reference/special-files/) and [special directories](https://www.chezmoi.io/reference/special-directories/) are meant to be created directly in `~/.local/share/chezmoi` (the source state/directory), and _not_ managed via `chezmoi add` (which would actually make sense, otherwise the home directory would get cluttered with `.chezmoi*` files and dirs). Is that correct?

**Reply by halostatue** [reply]
That is *mostly* correct. If you have a `.chezmoiroot` file, that must *always* be in the root of the working tree (the repo), and it must name a relative directory inside of the working tree. All of the *other* special files and directories must be in the source tree (either the working tree if you do not have `.chezmoiroot` or the directory pointed to if you do).

Note that *some* special files and directories may be repeated multiple times *within* the tree, and affect state processing for the paths therein.

**Reply by chuckwondo** [reply]
@halostatue, thanks so much for the additional details. This is all very helpful.
